Simplify logic to resolve tasks stuck in queued despite stalled_task_timeout

RNHTTR opened the pull request:
I accidentally closed #30108, so this is basically reopening that PR. Some updates:

* There's not really a mechanism to deprecate multiple configs into one. The nature of this change requires some unique deprecation logic which is implemented in `scheduler_job.py`.
* This deprecates `celery.stalled_task_timeout`, `kubernetes.worker_pods_pending_timeout`, and `celery.task_adoption_timeout`

closes: #28120
closes: #21225
closes: #28943

Tasks occasionally get stuck in queued and aren't resolved by `stalled_task_timeout` (#28120). This PR moves the logic for handling stalled tasks to the scheduler and simplifies the logic by marking any task that has been queued for more than `scheduler.task_queued_timeout` as failed, allowing it to be retried if the task has available retries.

This doesn't require an additional scheduler nor allow for the possibility of tasks to get stuck in an infinite loop of scheduled -> queued -> scheduled ... -> queued as exists in #28943.

RNHTTR committed:
simplify and consolidate logic for tasks stuck in queued

RNHTTR committed:
simplify and consolidate logic for tasks stuck in queued

RNHTTR committed:
simplify and consolidate logic for tasks stuck in queued

ephraimbuddy commented on airflow/jobs/scheduler_job.py:
```diff
@@ -152,6 +152,43 @@ def __init__(
+                "The 'stalled_task_timeout' parameter is deprecated. "
+                "Please use 'scheduler.task_queued_timeout'.",
+                RemovedInAirflow3Warning,
+                stacklevel=2,
```
Registering the deprecation at https://github.com/apache/airflow/blob/85427755f30d72e6c401ddc01b694b8a181c9a96/airflow/configuration.py#L175-L179 is enough. Airflow will automatically warn. No need to have another warning

RNHTTR replied:
Based on [this comment](https://github.com/apache/airflow/pull/30375#discussion_r1157632498), I'm not sure it would make sense to deprecate in this manner in case one of these timeouts is intentionally set longer than whatever default is chosen for `task_queued_timeout`

jedcunningham replied:
It might be worth iterating on this a bit, as ideally we'd get the old configs listed here so [this](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#base-log-folder-deprecated) still happens.

Maybe we get a "removed options" that doesn't do the "old first, then new" precedence when reading?

ephraimbuddy commented on airflow/jobs/scheduler_job.py:
```diff
@@ -152,6 +152,43 @@ def __init__(
+        task_queued_timeout = conf.getfloat("scheduler", "task_queued_timeout")
+        self._task_queued_timeout = max(
+            stalled_task_timeout, task_adoption_timeout, worker_pods_pending_timeout, task_queued_timeout
+        )
```
Why are we taking the max value? Won't it be better to have a single config value? Having a single point of truth is better in my opinion.

RNHTTR replied:
The idea to take the max is to ensure we won't step on the toes of anyone who, for some reason, has some timeout longer than the default for `task_queued_timeout`. cc @jedcunningham

jedcunningham replied:
Just to state it a little differently, we are merging stalled_task_timeout, task_adoption_timeout, and worker_pods_pending_timeout into a single new timeout, task_queued_timeout. In order to honor any existing changes to those now-removed timeouts, we will simply just run it at the highest value instead.

ephraimbuddy commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1408,6 +1443,40 @@ def _send_sla_callbacks_to_processor(self, dag: DAG) -> None:
+    @provide_session
+    def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
+        """
+        Mark tasks stuck in queued for longer than `task_queued_timeout` as failed.
```
In the doc string here, we are using `task_queued_timeout` but self._task_queued_timeout is a max of many other configs

ephraimbuddy commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1408,6 +1443,40 @@ def _send_sla_callbacks_to_processor(self, dag: DAG) -> None:
+        available retries, it will be retried.
+        """
+        self.log.debug("Calling SchedulerJob._fail_tasks_stuck_in_queued method")
+        try:
```
Instead of this try, I think we should use `run_with_db_retries` see `adopt_or_reset_orphaned_tasks` below

ephraimbuddy reviewed (commented):
Just reviewed only the scheduler_job, will come back again

RNHTTR committed:
fixed tests; updated fail stuck tasks to use run_with_db_retries

RNHTTR committed:
mypy; fixed tests

RNHTTR committed:
fix task_adoption_timeout in celery integration test

jedcunningham commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1408,6 +1443,51 @@ def _send_sla_callbacks_to_processor(self, dag: DAG) -> None:
+        for attempt in run_with_db_retries(logger=self.log):
+            with attempt:
+                self.log.debug(
+                    "Running SchedulerJob.adopt_or_reset_orphaned_tasks with retries. Try %d of %d",
```
```suggestion
                    "Running SchedulerJob._fail_tasks_stuck_in_queued with retries. Try %d of %d",
```

jedcunningham commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1408,6 +1443,51 @@ def _send_sla_callbacks_to_processor(self, dag: DAG) -> None:
+                    query = session.query(TI).filter(
+                        TI.state == State.QUEUED,
+                        TI.queued_dttm < (timezone.utcnow() - timedelta(seconds=self._task_queued_timeout)),
+                    )
```
Am I missing where we call `cleanup_stuck_queued_tasks`?

jedcunningham commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1408,6 +1443,51 @@ def _send_sla_callbacks_to_processor(self, dag: DAG) -> None:
+                            repr(ti),
+                            self._task_queued_timeout,
+                        )
+                        ti.handle_failure(error=msg, session=session)
```
I'm less familiar with celery, but with KE you don't have to do this explicitly, deleting the pod is enough. If we both fail it here and delete it, we should double check we aren't using up multiple retries.

ephraimbuddy replied:
When a task is marked as failed externally, failure callbacks are not called https://github.com/apache/airflow/blob/0b83f064e524272e8fc1dfcd38edd7ad39bc6660/airflow/models/taskinstance.py#L1468-L1470. 

My suggestion is that these task instances should not be marked failed here instead we can collect the task instances that have overstayed and send them to the executor and have the executor purge them from the running tasks. The failure would now be more natural when they come back to the scheduler. We would get a message like `the executor reports that the task instance failed but it's still queued`, that way failure callback and retries will still work. And I think that's the plan here but don't know why we called the failure callback

jedcunningham commented on tests/jobs/test_scheduler_job.py:
```diff
@@ -1531,6 +1531,50 @@ def test_adopt_or_reset_orphaned_tasks(self, dag_maker):
+        self.scheduler_job.processor_agent = processor
+        self.scheduler_job._task_queued_timeout = 300
+
+        self.scheduler_job._fail_tasks_stuck_in_queued()
```
Should we mock `cleanup_stuck_queued_tasks` and make sure it is called with the TI?

RNHTTR committed:
addressing comments

ephraimbuddy commented on airflow/executors/celery_executor.py:
```diff
@@ -414,23 +414,23 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
-                    app.control.revoke(celery_async_result.task_id)
-                except Exception as ex:
-                    self.log.error("Error revoking task instance %s from celery: %s", task_instance_key, ex)
+        print(f"\n\nti: {type(ti)}\n\n")
```
```suggestion
```
I don't think we need this

ephraimbuddy commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1478,6 +1478,7 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
                             self._task_queued_timeout,
                         )
                         ti.handle_failure(error=msg, session=session)
+                        self.executor.cleanup_stuck_queued_task(ti)
```
Do we really need the ti.handle_failure? Won't it be handled in `Scheduler._process_executor_events`?

Also, I think we shouldn't iterate the task instances here but send all to the executor, then have the executor iterate and remove them from the running. When we iterate here, we call the executor cleanup several times but we could call it once with the task instances and then have it remove them from the running tasks. Looks more efficient _I think_

RNHTTR committed:
remove useless print

RNHTTR committed:
fix typo

RNHTTR committed:
move failure logic to executor

RNHTTR committed:
fix scheduler job test

ephraimbuddy commented on airflow/executors/celery_executor.py:
```diff
@@ -543,22 +417,28 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
+        readable_tis = []
+        for ti in tis:
+            readable_tis.append(repr(ti))
+            task_instance_key = ti.key
```
```suggestion
            task_instance_key = ti.key
            self.running.discard(task_instance_key)
```
Should we also remove it from running?

ephraimbuddy replied:
Removing it is like failing it/sending it back to scheduler.

RNHTTR replied:
I instead decided to call `self.fail(...)` which ultimately calls `BaseExecutor.change_state(...)` which removes the key from running. I feel like that's more explicit... what do you think?

ephraimbuddy replied:
Removing it from running seems better since the task would be sent back to scheduler. The executor will report that the task has finished

RNHTTR replied:
What do you mean by sent back to the scheduler?

jedcunningham replied:
Okay, so I had a chat with @ephraimbuddy and I think we should give this a try instead. Let's remove it from running, and add the failure to the event_buffer explicitly. In theory, then we don't have to set the task as failed here.

This is essentially what KE is doing indirectly.

ephraimbuddy replied:
Sorry, I think your code is right. Failing it adds it to event buffer but just removing it doesn't and I think that could be the bug in the previous fix with the stalled task timeout since it didn't put the task in the event buffer

RNHTTR committed:
adjustments for new scheduler job

RNHTTR committed:
appeasing static checks

RNHTTR committed:
fix test for new scheduler job paradigm

RNHTTR committed:
Updating docs for deprecations

jedcunningham commented on airflow/configuration.py:
```diff
@@ -236,6 +236,16 @@ class AirflowConfigParser(ConfigParser):
     }
 
+    # A mapping of new configurations to a list of old configurations for when one configuration
+    # deprecates more than one other deprecation.
```
```suggestion
    # deprecates more than one other deprecation.
    # There is no special handling for values set under the old options - you must handle that yourself.
```

Or something like that. Let's be explicit.

jedcunningham commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -854,6 +815,45 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
+        """
+        readable_tis = []
+        for ti in tis:
+            readable_tis.append(repr(ti))
```
```suggestion
```

Nit: these next few are just refactoring how this is built to be simpler.

1 of 4

jedcunningham commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -854,6 +815,45 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
+            ).items
+            if not pod_list:
+                # Remove from list of tis that were cleaned up
+                readable_tis.pop()
```
```suggestion
```

2 of 4

jedcunningham commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -854,6 +815,45 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
+                continue
+            elif len(pod_list) > 1:
+                # Remove from list of tis that were cleaned up
+                readable_tis.pop()
```
```suggestion
```

3 of 4

jedcunningham commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -854,6 +815,45 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
+                readable_tis.pop()
+                self.log.warning("Found multiple pods for ti %s: %s", ti, pod_list)
+                continue
+            self.kube_scheduler.delete_pod(pod_id=pod_list[0].metadata.name, namespace=namespace)
```
```suggestion
            readable_tis.append(repr(ti))
            self.kube_scheduler.delete_pod(pod_id=pod_list[0].metadata.name, namespace=namespace)
```

4 of 4

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1427,6 +1463,53 @@ def _send_sla_callbacks_to_processor(self, dag: DAG) -> None:
+        `celery.stalled_task_timeout`, or `celery.task_adoption_timeout` is set and
+        has a higher value than `scheduler.task_queued_timeout`, that value will be
+        used as `task_queued_timeout`.
+
```
```suggestion
```

Not sure we need this comment here, this is handled in a different scope.

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1427,6 +1463,53 @@ def _send_sla_callbacks_to_processor(self, dag: DAG) -> None:
+                    tasks_stuck_in_queued: list[TaskInstance] = with_row_locks(
+                        query, of=TI, session=session, **skip_locked(session=session)
+                    ).all()
+                    print(f"\n\ntis: {tasks_stuck_in_queued}\n\n")
```
```suggestion
                    print(f"\n\ntis: {tasks_stuck_in_queued}\n\n")
```

Leftover debugging? Or, switch this to debug or warning log instead (with some extra context).

RNHTTR committed:
news & small changes

RNHTTR committed:
news & small changes

jedcunningham commented on newsfragments/30375.significant.rst:
```diff
@@ -0,0 +1,9 @@
+Simplify logic to resolve tasks stuck in queued despite stalled_task_timeout
```
```suggestion
Consolidate handling of tasks stuck in queued under new ``task_queued_timeout`` config
```

This is probably a better title?

jedcunningham commented on newsfragments/30375.significant.rst:
```diff
@@ -0,0 +1,9 @@
+`kubernetes.worker_pods_pending_timeout`, `celery.stalled_task_timeout`, and 
+`celery.task_adoption_timeout`. If any of these configurations are set, the longest timeout will be
+respected. For example, if `celery.stalled_task_timeout` is 1200, and `scheduler.task_queued_timeout` 
+is 600, Airflow will set `scheduler.task_queued_timeout` to 1200.
```
```suggestion
Logic for handling tasks stuck in the queued state has been consolidated, and the all configurations
responsible for timing out stuck queued tasks have been deprecated and merged into
``scheduler.task_queued_timeout``. The configurations that have been deprecated are
``kubernetes.worker_pods_pending_timeout``, ``celery.stalled_task_timeout``, and
``celery.task_adoption_timeout``. If any of these configurations are set, the longest timeout will be
respected. For example, if ``celery.stalled_task_timeout`` is 1200, and ``scheduler.task_queued_timeout``
is 600, Airflow will set ``scheduler.task_queued_timeout`` to 1200.

```

Fixes static checks.

jedcunningham commented on newsfragments/30375.significant.rst:
```diff
@@ -0,0 +1,9 @@
+
+Logic for handling tasks stuck in the queued state has been consolidated, and the all configurations 
+responsible for timing out stuck queued tasks have been deprecated and merged into 
+`scheduler.task_queued_timeout`. The configurations that have been deprecated are 
```
```suggestion
`[scheduler] task_queued_timeout`. The configurations that have been deprecated are 
```

Let's reference these config options like this instead.

jedcunningham commented on airflow/executors/base_executor.py:
```diff
@@ -376,6 +376,18 @@ def terminate(self):
+        :param tis: List of Task Instances to clean up
+        :return: List of readable task instances for a warning message
+        """
+        raise NotImplementedError()
```
I just realized we haven't properly handled what happens when another executor hits this (however unlikely).

Easy way to reproduce this would be setting a really low task_queued_timeout, and add a longer sleep to the top level of a DAG, when using LocalExecutor.

Maybe log this should just call fail also? It shouldn't kill the scheduler I'd think.

RNHTTR committed:
Update newsfragments/30375.significant.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

RNHTTR committed:
Update newsfragments/30375.significant.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

RNHTTR committed:
added cleanup stuck task functionality to base executor

jedcunningham commented on airflow/executors/base_executor.py:
```diff
@@ -297,8 +297,10 @@ def change_state(self, key: TaskInstanceKey, state: str, info=None) -> None:
         """
         self.log.debug("Changing state: %s", key)
         try:
+            print(f"\n\n removing key {key}")
```
```suggestion
```

jedcunningham commented on airflow/executors/base_executor.py:
```diff
@@ -297,8 +297,10 @@ def change_state(self, key: TaskInstanceKey, state: str, info=None) -> None:
+            print(f"\n\n removing key {key}")
             self.running.remove(key)
         except KeyError:
+            print(f"Key error!")
```
```suggestion
```

jedcunningham commented on airflow/executors/base_executor.py:
```diff
@@ -376,6 +378,22 @@ def terminate(self):
         """This method is called when the daemon receives a SIGTERM."""
         raise NotImplementedError()
 
+    def cleanup_stuck_queued_tasks(self, tis: list[TaskInstance]) -> list[str]:  # pragma: no cover
```
```suggestion
    def cleanup_stuck_queued_tasks(self, tis: list[TaskInstance]) -> list[str]:
```

jedcunningham commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -865,6 +826,41 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
+
+        :param tis: List of Task Instances to clean up
+        :return: List of readable task instances for a warning message
+        """
```
```suggestion
        """
        if TYPE_CHECKING:
            assert self.kube_client
            assert self.kube_scheduler
            
```

This should fix mypy.

RNHTTR committed:
fix sloppy mistakes & mypy

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1425,6 +1461,47 @@ def _send_sla_callbacks_to_processor(self, dag: DAG) -> None:
+                try:
+                    query = session.query(TI).filter(
+                        TI.state == State.QUEUED,
+                        TI.queued_dttm < (timezone.utcnow() - timedelta(seconds=self._task_queued_timeout)),
```
```suggestion
                        TI.queued_dttm < (timezone.utcnow() - timedelta(seconds=self._task_queued_timeout)),
                        TI.queued_by_job_id == self.job_id,
```

Lets only look at this specific schedulers tasks.

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1425,6 +1461,47 @@ def _send_sla_callbacks_to_processor(self, dag: DAG) -> None:
+                    )
+                    tasks_stuck_in_queued: list[TaskInstance] = with_row_locks(
+                        query, of=TI, session=session, **skip_locked(session=session)
+                    ).all()
```
```suggestion
                    tasks_stuck_in_queued: list[TaskInstance] = query.all()
```

Which also means we don't need to lock rows.

RNHTTR committed:
removing self.fail from base_executor

RNHTTR committed:
Update airflow/jobs/scheduler_job_runner.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

RNHTTR committed:
Update airflow/jobs/scheduler_job_runner.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

jedcunningham committed:
Fix job_id filter

jedcunningham committed:
Don't even run query if executor doesn't support timing out queued tasks

jedcunningham committed:
Add support for LocalKubernetesExecutor and CeleryKubernetesExecutor

jedcunningham committed:
Add config option to control how often it runs - we want it quicker than
the timeout

jedcunningham committed:
Fixup newsfragment

jedcunningham committed:
mark old KE pending pod check interval as deprecated by new check interval

jedcunningham committed:
Fixup deprecation warnings

This more closely mirrors how deprecations are raised for "normal"
deprecations.

I've removed the depth, as moving up the stack doesn't really help the
user at all in this situation.

jedcunningham committed:
Another deprecation cleanup

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1425,6 +1465,51 @@ def _send_sla_callbacks_to_processor(self, dag: DAG) -> None:
+                    "Running SchedulerJob._fail_tasks_stuck_in_queued with retries. Try %d of %d",
+                    attempt.retry_state.attempt_number,
+                    MAX_DB_RETRIES,
+                )
```
I'm really not sure we need to use retries on this?

jedcunningham replied:
I removed this.

jedcunningham committed:
Remove db retries

ephraimbuddy commented on tests/jobs/test_scheduler_job.py:
```diff
@@ -1572,6 +1572,26 @@ def test_adopt_or_reset_orphaned_tasks(self, dag_maker):
         assert ti2.state == State.QUEUED, "Tasks run by Backfill Jobs should not be reset"
 
+    def test_fail_stuck_queued_tasks(self, dag_maker):
+        session = settings.Session()
```
```suggestion
    def test_fail_stuck_queued_tasks(self, dag_maker, session):
```

ephraimbuddy committed:
Fix test

hterik commented on airflow/executors/celery_executor.py:
```diff
@@ -543,22 +417,28 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
+        for ti in tis:
+            readable_tis.append(repr(ti))
+            task_instance_key = ti.key
+            self.fail(task_instance_key, None)
```
Not sure if the data model allows, but is it possible to add some error details to the taskinstance here so a user can understands why it failed, by just looking in the web ui? 

To avoid admins getting asked a lot of questions from users like "why did this task fail without any logs", having to open up scheduler logs.

RNHTTR replied:
My understanding is that there's not an easy way to surface these logs in the UI. Such "missing" task logs could also be caused by [zombies](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#zombie-undead-tasks), which can be caused by... tons of stuff.

I think a good intermediate step will be to add a blurb in the docs about missing task logs for zombies as well as tasks stuck in queued. I plan to open such a docs PR soon-ish.

RNHTTR replied:
This will likely be possible in Airflow 2.8.0 with this PR: https://github.com/apache/airflow/pull/32646
